turbopack/crates/turbo-tasks-backend/src/utils/dash_map_multi.rs

```diff
@@ -0,0 +1,178 @@
+use std::{
+    hash::{BuildHasher, Hash},
+    ops::{Deref, DerefMut},
+    sync::Arc,
+};
+
+use dashmap::{DashMap, RwLockWriteGuard, SharedValue};
+use hashbrown::{hash_map, HashMap};
+
+pub enum RefMut<'a, K, V, S>
+where
+    S: BuildHasher,
+{
+    Base(dashmap::mapref::one::RefMut<'a, K, V, S>),
+    Simple {
+        #[allow(dead_code)]
+        guard: RwLockWriteGuard<'a, HashMap<K, SharedValue<V>, S>>,
+        key: *const K,
+        value: *mut V,
+    },
+    Shared {
+        #[allow(dead_code)]
+        guard: Arc<RwLockWriteGuard<'a, HashMap<K, SharedValue<V>, S>>>,
+        key: *const K,
+        value: *mut V,
+    },
+}
+
+unsafe impl<'a, K: Eq + Hash + Sync, V: Sync, S: BuildHasher> Send for RefMut<'a, K, V, S> {}
+unsafe impl<'a, K: Eq + Hash + Sync, V: Sync, S: BuildHasher> Sync for RefMut<'a, K, V, S> {}
+
+impl<'a, K: Eq + Hash, V, S: BuildHasher> RefMut<'a, K, V, S> {
+    pub fn key(&self) -> &K {
+        self.pair().0
+    }
+
+    pub fn value(&self) -> &V {
+        self.pair().1
+    }
+
+    pub fn value_mut(&mut self) -> &mut V {
+        self.pair_mut().1
+    }
+
+    pub fn pair(&self) -> (&K, &V) {
+        match self {
+            RefMut::Base(r) => r.pair(),
+            &RefMut::Simple { key, value, .. } => unsafe { (&*key, &*value) },
+            &RefMut::Shared { key, value, .. } => unsafe { (&*key, &*value) },
+        }
+    }
+
+    pub fn pair_mut(&mut self) -> (&K, &mut V) {
+        match self {
+            RefMut::Base(r) => r.pair_mut(),
+            &mut RefMut::Simple { key, value, .. } => unsafe { (&*key, &mut *value) },
+            &mut RefMut::Shared { key, value, .. } => unsafe { (&*key, &mut *value) },
+        }
+    }
+}
+
+impl<'a, K: Eq + Hash, V, S: BuildHasher> Deref for RefMut<'a, K, V, S> {
+    type Target = V;
+
+    fn deref(&self) -> &V {
+        self.value()
+    }
+}
+
+impl<'a, K: Eq + Hash, V, S: BuildHasher> DerefMut for RefMut<'a, K, V, S> {
+    fn deref_mut(&mut self) -> &mut V {
+        self.value_mut()
+    }
+}
+
+impl<'a, K, V, S> From<dashmap::mapref::one::RefMut<'a, K, V, S>> for RefMut<'a, K, V, S>
+where
+    K: Hash + Eq,
+    S: BuildHasher,
+{
+    fn from(r: dashmap::mapref::one::RefMut<'a, K, V, S>) -> Self {
+        RefMut::Base(r)
+    }
+}
+
+pub fn get_multiple_mut<K, V, S>(
+    map: &DashMap<K, V, S>,
+    key1: K,
+    key2: K,
+    insert_with: impl Fn() -> V,
+) -> (RefMut<'_, K, V, S>, RefMut<'_, K, V, S>)
+where
+    K: Hash + Eq + Clone,
+    S: BuildHasher + Clone,
+{
```

Comment:
As this is a risk for deadlocks, this should have a unit test that stress-tests this on a `DashMap` with a small number of shards.